src/utils/helpers/revalidateAll.tsx

```diff
@@ -0,0 +1,73 @@
+import {
+  ApplicationElementStates,
+  ResponsesByCode,
+  RevalidateResult,
+  ValidityFailure,
+} from '../types'
+import { defaultValidate } from '../../formElementPlugins/defaultValidate'
+
+export const revalidateAll = async (
+  elementsState: ApplicationElementStates | undefined,
+  responsesByCode: ResponsesByCode | undefined,
+  strict = true,
+  shouldUpdateDatabase = true
+): Promise<RevalidateResult> => {
+  const validate = defaultValidate // To-Do: import custom validation methods
+
+  // console.log('Elements', elementsState)
+  // console.log('responses', responsesByCode)
+  if (elementsState) {
+    const elementCodes = Object.keys(elementsState).filter(
+      (key) =>
+        responsesByCode && // Typescript requires this
+        elementsState[key].category === 'QUESTION' &&
+        elementsState[key].isVisible === true &&
+        // Strict/Loose validation logic:
+        ((strict && elementsState[key].isRequired) ||
+          (strict && responsesByCode[key]?.isValid !== null) ||
+          (!strict && responsesByCode[key]?.isValid !== null))
+    )
+
+    const validationExpressions = elementCodes.map((code) => elementsState[code].validation)
+
+    const evaluatedValidations = []
+
+    for (let i = 0; i < elementCodes.length; i++) {
+      const code = elementCodes[i]
+      const thisResponse = responsesByCode
+        ? responsesByCode[code]?.text
+          ? responsesByCode[code]?.text
+          : ''
+        : ''
+      const responses = { thisResponse, ...responsesByCode }
+      const expression = validationExpressions[i]
+      evaluatedValidations.push(
+        validate(expression, elementsState[code]?.validationMessage as string, {
+          objects: [responses],
+          APIfetch: fetch,
+        })
+      )
+    }
+    const resultArray = await Promise.all(evaluatedValidations)
+    // console.log('Result', resultArray)
+    const validityFailures: ValidityFailure[] = []
+    if (shouldUpdateDatabase) {
+      elementCodes.forEach((code, index) => {
+        if (!resultArray[index].isValid) {
+          validityFailures.push({
+            id: (responsesByCode && responsesByCode[code].id) || 0,
+            isValid: false,
+            code,
+          })
+        }
+      })
+    }
+
+    return {
+      allValid: resultArray.every((element) => element.isValid),
+      validityFailures,
+    }
+  }
+  // Typescript requires final return -- it should never be reached
+  return { allValid: false, validityFailures: [] }
```

Comment:
This is just because `elementState` can be undefined... it might be the case it reaches this point. Should change to not allow receiving elementState as undefined?

Reply:
Yeah, I couldn't get it to work. The `elementState` object should never be undefined by the time it gets to this function, but it _can_ be undefined elsewhere (like in `useGetResponsesAndElementState` for example), so I wasn't able to insist that this function take `elementState` that is never undefined. Any ideas?

Reply:
You can type to the `ApplicationElementState` if you are sure it would only reach the call in case it was defined.
Example:
```
    const revalidate = await revalidateAll(
      elementsState as ApplicationElementStates,
      responsesByCode
    )
```
And change the `revalidateAll` to expect only the defined object

Reply:
Got it! Thanks for clarifying :) 179cbdc